### PR TITLE
Use hash.BucketSet for StatefulSet mode

### PR DESCRIPTION
**Pull request by yanweiguo**
1. Use `hash.BucketSet` for StatefulSet mode to build leader elector.
1. Introduce `NewStatefulSetBucketAndSet` func to wrap information for a StatefulSet pod to create the `Bucket` and `BucketSet`.
1. Introduce `BucketList` func for `BucketSet` to get all candidate bucket names. (with this we can get rid of `ControllerOrdinal`)
1. Remove `ControllerOrdinal` func.
1. Add direct unit test for `newStatefulSetConfig`.

Downstream change: https://github.com/knative/serving/pull/8581

/assign @vagababov @mattmoor

### CODE REVIEW

File: leaderelection/context_test.go [hunk 1]
```diff
@@ -146,12 +147,48 @@ func TestWithBuilder(t *testing.T) {
...
 
+func TestNewStatefulSetBucketAndSet(t *testing.T) {
+	wantNames := []string{
+		"http://as-0.autoscaler.knative-testing.svc.cluster.local:80",
+		"http://as-1.autoscaler.knative-testing.svc.cluster.local:80",
+		"http://as-2.autoscaler.knative-testing.svc.cluster.local:80",
+	}
+
+	os.Setenv(controllerOrdinalEnv, "as-2")
+	defer os.Unsetenv(controllerOrdinalEnv)
```
Comment (vagababov) [inline leaderelection/context_test.go:158]: t.Cleanup?

File: leaderelection/context_test.go [hunk 2]
```diff
@@ -146,12 +147,48 @@ func TestWithBuilder(t *testing.T) {
...
+	os.Setenv(serviceNameEnv, "autoscaler")
+	defer os.Unsetenv(serviceNameEnv)
+
+	_, _, err := NewStatefulSetBucketAndSet(2)
+	if err == nil {
+		// Ordinal 2 should be range [0, 2)
+		t.Fatal("expected error from NewStatefulSetBucketAndSet but got nil")
```
Comment (vagababov) [inline leaderelection/context_test.go:165]:
```suggestion
		t.Fatal("Expected error from NewStatefulSetBucketAndSet but got nil")
```

File: leaderelection/context_test.go [hunk 3]
```diff
@@ -146,12 +147,48 @@ func TestWithBuilder(t *testing.T) {
...
+	}
+
+	bkt, bs, err := NewStatefulSetBucketAndSet(3)
+	if err != nil {
+		// Ordinal 2 should be range [0, 2)
+		t.Fatalf("NewStatefulSetBucketAndSet() = %d", err)
```
Comment (vagababov) [inline leaderelection/context_test.go:171]:
```suggestion
		t.Fatal("NewStatefulSetBucketAndSet() =", err)
```

**Comment by vagababov** [comment]
We probably also need how to propagate the BucketSet down.
Attach to context?

**Comment by yanweiguo** [comment]
> We probably also need how to propagate the BucketSet down.
> Attach to context?

I plan to use `NewStatefulSetBucketAndSet` in downstream to get the `Bucket` (for checking whether a pod has the key) and `BucketSet` (for forwarding).

File: hash/bucketer.go [hunk 4]
```diff
@@ -99,6 +100,14 @@ func (b *BucketSet) Owner(key string) string {
 	return ret
 }
 
+// BucketList returns the bucket names of this BucketSet in random order.
+func (b *BucketSet) BucketList() []string {
```
Comment (vagababov) [inline hash/bucketer.go:104]: Is this for testing only? Or it will be used for production code as well?

Comment (yanweiguo) [reply hash/bucketer.go:104]: In production code to build websocket connections.

File: leaderelection/context.go [hunk 5]
```diff
@@ -173,29 +175,50 @@ func (b *standardBuilder) buildElector(ctx context.Context, la reconciler.Leader
...
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if ssc.StatefulSetID.ordinal >= buckets {
+		return nil, nil, fmt.Errorf("ordinal %d is out of range [0, %d)",
+			ssc.StatefulSetID.ordinal, buckets)
+	}
+
+	names := sets.String{}
```
Comment (mattmoor) [inline leaderelection/context.go:207]: Does this support `make(sets.String, buckets)` since it's just a type alias of map? 🤔

**Comment by mattmoor** [comment]
Why wouldn't we use this for all of our bucketing?

File: leaderelection/context.go (hunk 5 above)
Comment (yanweiguo) [reply leaderelection/context.go]: yes. Change.

**Comment by yanweiguo** [comment]
> Why wouldn't we use this for all of our bucketing?

I have that in mind. I prefer to do it for standard mode in a separate PR (smaller changes).

**Pull request merged**